ipld/kamt/src/kamt.rs

```diff
@@ -352,6 +364,133 @@ where
         V: DeserializeOwned,
         F: FnMut(&K, &V) -> anyhow::Result<()>,
     {
-        self.root.for_each(self.store.borrow(), &mut f)
+        for res in self {
+            let (k, v) = res?;
+            (f)(k, v)?;
+        }
+        Ok(())
+    }
+}
+
+impl<BS, V, K, H, const N: usize> Kamt<BS, K, V, H, N>
+where
+    K: DeserializeOwned + PartialOrd,
+    V: DeserializeOwned,
+    BS: Blockstore,
+{
+    /// Returns an iterator over the entries of the map.
+    ///
+    /// The iterator element type is `Result<(&'a K, &'a V), Error>`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fvm_ipld_kamt::Kamt;
+    /// use fvm_ipld_kamt::id::Identity;
+    ///
+    /// let store = fvm_ipld_blockstore::MemoryBlockstore::default();
+    ///
+    /// let mut map: Kamt<_, u32, _, Identity> = Kamt::new(store);
+    /// map.set(1, 1).unwrap();
+    /// map.set(4, 2).unwrap();
+    ///
+    /// let mut x: u32 = 0;
+    /// for res in map.iter() {
+    ///     let (key, value) = res.unwrap();
+    ///     println!("key: {}, value: {}", key, value);
+    ///     x = x+1;
+    /// }
+    /// assert_eq!(x,2)
+    /// ```
+    pub fn iter(&self) -> IterImpl<BS, V, K, H, N> {
+        IterImpl::new(&self.store, &self.root)
+    }
+
+    /// Iterate over the KAMT starting at the given key.
+    /// iteration:
+    /// # Examples
+    ///
+    /// ```
+    /// use fvm_ipld_kamt::Kamt;
+    /// use fvm_ipld_kamt::id::Identity;
+    ///
+    /// let store = fvm_ipld_blockstore::MemoryBlockstore::default();
+    ///
+    /// let mut map: Kamt<_, u32, _, Identity> = Kamt::new(store);
+    /// map.set(1, 1).unwrap();
+    /// map.set(2, 4).unwrap();
+    /// map.set(3, 3).unwrap();
+    /// map.set(4, 2).unwrap();
+    ///
+    /// let mut results = map.iter().take(2).collect::<Result<Vec<_>, _>>().unwrap();
+    ///
+    /// let last_key = results.last().unwrap().0;
+    ///
+    /// for res in map.iter_from(last_key).unwrap().skip(1) {
+    ///     results.push(res.unwrap());
+    /// }
+    ///
+    /// println!("{:?}", results);
+    /// assert_eq!(results.len(), 4);
+    /// ```
+
+    /// Iterate over the KAMT starting at the given key. This can be used to implement "ranged" iteration:
+    ///
+    /// ```rust
+    /// use fvm_ipld_kamt::Kamt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    /// use fvm_ipld_kamt::id::Identity;
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// // Create a Kamt with 5 keys, a-e.
+    /// let mut kamt: Kamt<_, u32, String, Identity> = Kamt::new_with_bit_width(store, 5);
```

Comment:
I'd just change this to `Kamt::new_with_config` (or drop the bit-width, I don't think that it's actually necessary to specify it?